Mfold

evelinabis opened the pull request.

evelinabis committed:
Mfold

evelinabis committed:
Mfold. Added comment task description

DmitriiSukhomlinov commented on src/corelibs/U2Gui/src/util/RegionSelector.h:
```diff
@@ -71,6 +71,8 @@ class U2GUI_EXPORT RegionSelector : public QWidget {
 
     void showErrorMessage();
 
+    QPair<QLineEdit*, QLineEdit*> getLineEdits() const;
```
Из-за этой функции поля, фактически, стали публичными, т.к. с помощью одного только указателя теперь вполне законно можно вертеть селектором как угодно извне, что не очень удачно с точки зрения инкапсуляции. Лучше убрать эту функцию и вместо неё завести пару сигналов `si_startEditChanged(const QString &text)` и `si_endEditChanged(const QString &text)`, которые прикрепить к сигналу `QLineEdit::textChanged(...)` каждого из полей соответственно, типа такого:

`connect(startEdit, &QLineEdit::textChanged, this, &RegionSelector::si_startEditChanged);`

И из вне селектора уже прицепиться к `RegionSelector::si_startEditChanged`. Т.е. сигнал будет проксироваться через `RegionSelector`

DmitriiSukhomlinov replied:
А вообще, заводить новые сигналы даже не нужно, т.к. уже существует `si_regionChanged()`, буквально двумя строчками ниже, не заметил его сразу

evelinabis replied:
What does this signal send if the region is invalid, for example, the end is not specified? In this case, how can we give a clear error to the user about what exactly he entered incorrectly?

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/ClickableLabel.cpp:
```diff
@@ -0,0 +1,13 @@
+// https://wiki.qt.io/Clickable_QLabel
```
Все новые файлы имеют нашу лицензию несмотря ни на что, но в файле нужно указать, где он заимствован на отдельной строке

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,173 @@
+            checkSave();
+        }
+    });
+    // todo use PathLineEdit?
```
Я не очень понял, что такое `PathLineEdit`, но в UGENE есть класс `SaveDocumentController`, который умеет задавать все необходимые параметры для подобных случаев.

evelinabis replied:
I couldn't find a way to make the output folder rather than a file.

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,173 @@
+    auto startVal = startEdit->text().toULongLong(&startOk);
+    auto endVal = endEdit->text().toULongLong(&endOk);
+
+    // Returns region len taking into account that the sequence can be circular. Doesn't validate region.
```
taken

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,173 @@
+    auto endVal = endEdit->text().toULongLong(&endOk);
+
+    // Returns region len taking into account that the sequence can be circular. Doesn't validate region.
+    auto getRegionLen = [this](qulonglong start, qulonglong end) {
```
Обычно, мы используем `qint64`. В любом случае, стоит использовать сигнал, про который я писал выше, и тебе не придется самой вычислять это значение

yalgaer replied:
Да, нужно поменять на qint64

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,173 @@
+    ui.wrongRegionLabel->setText(err);
+}
+
+void MfoldDialog::checkNotSave() {
```
Лучше сделать одну функцию с аргументом вместо двух `checkNotSave` и `checkSave`

evelinabis replied:
Agree

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,173 @@
+                                                                   tr("Invalid out path!"),
+                                                                   QMessageBox::Ok,
+                                                                   this);
+        msgBox->setInformativeText(tr("Output dir is read-only"));
```
Есть стандартное сообщение об ошибках `L10N::errorOpeningFileWrite()`

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.h:
```diff
@@ -0,0 +1,75 @@
+
+// Dialog for Mfold action.
+class MfoldDialog final : public QDialog {
+    Q_OBJECT
```
Я вот незнаю, замечание это или нет, но, обычно, сначала идут публичные поля, потом protected и потом уже приватные. Это, конечно, больше выглядит как вопрос стиля, но, мне все-таки кажется, что его стоит придерживаться

evelinabis replied:
Agree

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.h:
```diff
@@ -0,0 +1,75 @@
+    QComboBox* comboBox = nullptr;
+
+    // To check the correctness of a region.
+    qint64 seqLen {};
```
Ты ничем не проинициализировала переменные. И лучше инициализировать их просто с помощью `=`, если это можно, т.к. инициализирующие фигурные скобки имеют очень много нюансов

evelinabis replied:
Offtop: Какие нюансы?
Ontop: I don't agree, but I'll do it.

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldTask.cpp:
```diff
@@ -0,0 +1,254 @@
+}
+
+void MfoldTask::prepare() {
+    CHECK(!hasError(), );
```
`CHECK_OP(stateInfo, );` в данном случае подходит лучше - она проверяет так же не была ли таска отменена пользователем

DmitriiSukhomlinov replied:
В других местах аналогично

evelinabis replied:
I agree, good point.

DmitriiSukhomlinov commented on src/plugins/external_tool_support/src/mfold/MfoldTask.cpp:
```diff
@@ -0,0 +1,254 @@
+    };
+
+    // Read input.
+    QString hairpinHtml = readFileInput("inp.txt.det.html");  // html with hairpin info tables
```
После этой строчки тоже стоит проверить `stateInfo`

yalgaer reviewed (changes requested):
Мои замечания:

1. 
Я попробовала запустить mfold из UGENE локально и каждый раз при переключении на отчет mfold у меня UGENE подвисает:
![image](https://github.com/ugeneunipro/ugene/assets/17791039/6a38d3e5-50e1-4ba1-b2bd-97e71951b731)

Секунд через 20 UGENE оживает снова, но только до тех пор пока я снова не переключаюсь на другое окно.

2.
Само представление отчета тоже может быть существенно улучшено:
Сейчас не видно структуру отчета - мне нужно скролить десятки страниц таблиц и картинок.
Отчет нужно сделать более структурным и понятным. 
Чтобы он не тормозил его можно либо сразу открывать во внешнем браузере и использовать полноценно JavaScript для этого, либо сделать в UGENE native поддержку для отчетов mfold которая не тормозит.

3.
Картинки, которые производит mfold - векторные. Отчете же они сжаты до вида, когда их сложно читать. Нужно как-то улучшить их отображение: сделать их читаемыми и как-то связанными с множеством табличных структур описанных в таблице которая дается рядом с картинкой.

4. Должна быть возможность вернуться к однажды полученному отчету - открыть его заново.  Сегодня если я закрою окно с результатами расчёта я не смогу вернуться к нему снова

5. Нужно подумать как отобразить разметку ДНК полученную в отчете на последовательность в виде аннотаций к анализируемой последовательности.  Иначе запуск тула mfold из UGENE уводит пользователя из UGENE - результаты никак (кроме внешнего HTML отчета) не представлены в модели данных  UGENE.

По самому коду замечаний которые нельзя было бы исправить за день-два нет (так как патч совсем небольшой)

evelinabis commented:
> Мои замечания:
> 
> 
> Я попробовала запустить mfold из UGENE локально и каждый раз при переключении на отчет mfold у меня UGENE подвисает: ![image](https://private-user-images.githubusercontent.com/17791039/302064284-6a38d3e5-50e1-4ba1-b2bd-97e71951b731.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcxMDUxODAsIm5iZiI6MTcwNzEwNDg4MCwicGF0aCI6Ii8xNzc5MTAzOS8zMDIwNjQyODQtNmEzOGQzZTUtNTBlMS00YmExLWIyYmQtOTdlNzE5NTFiNzMxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA1VDAzNDgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmYmFiNzczOTY5ZDdiY2QxYjJhNTNhM2JhMTk0YWViOTBmODBjZjVlNGU0ZTRjM2IzN2Q4ZmE1OWE4OWRjMTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LrG4jJisysmZwgR59r7TdC-yRjAjqAEi-SgaAdnvZLs)
> 
> Секунд через 20 UGENE оживает снова, но только до тех пор пока я снова не переключаюсь на другое окно.

Agree. The problem here is the huge nested tables.

We can either display only pictures, or not display a report at all, or limit the length of the sequence. It makes more sense to display only pictures if there are a lot of structures.

> Само представление отчета тоже может быть существенно улучшено: Сейчас не видно структуру отчета - мне нужно скролить десятки страниц таблиц и картинок. Отчет нужно сделать более структурным и понятным. Чтобы он не тормозил его можно либо сразу открывать во внешнем браузере и использовать полноценно JavaScript для этого, либо сделать в UGENE native поддержку для отчетов mfold которая не тормозит.

Great wish! I like the idea of JS+opening in an external browser.
If we do this, where will the "Open in browser" button be?

Also, how do you see native support in UGENE? Add a window to SV displaying an image without a table?

> Картинки, которые производит mfold - векторные. Отчете же они сжаты до вида, когда их сложно читать. Нужно как-то улучшить их отображение: сделать их читаемыми и как-то связанными с множеством табличных структур описанных в таблице которая дается рядом с картинкой.

Regarding the size of the picture: I agree that changes are required, I wrote them in a separate message, they must be made. But I don’t understand what you mean by "связанными с множеством табличных структур".

> 4. Должна быть возможность вернуться к однажды полученному отчету - открыть его заново.  Сегодня если я закрою окно с результатами расчёта я не смогу вернуться к нему снова

Always save output files, like in the workflow? And the dashboard is like in the workflow?)

> 5. Нужно подумать как отобразить разметку ДНК полученную в отчете на последовательность в виде аннотаций к анализируемой последовательности.  Иначе запуск тула mfold из UGENE уводит пользователя из UGENE - результаты никак (кроме внешнего HTML отчета) не представлены в модели данных  UGENE.

Okay, I'll do it.

To sum up the wishes: we need to do a lot, cool and well. I agree that I would like this, but how long it will take needs to be discussed.

evelinabis commented:
Closed for revision. New comments and continued discussions are welcome. Thanks for your comments.

yalgaer commented:
> Great wish! I like the idea of JS+opening in an external browser. If we do this, where will the "Open in browser" button be?

Если мы идем путем работы с внешним браузером, то сам отчет о выполнении расчёта может содержать краткую выжимку и внешнюю ссылку, которую и откроет внешний браузер.

yalgaer commented:
> Also, how do you see native support in UGENE? Add a window to SV displaying an image without a table?

Внутренние средства визуализации UGENE данных поверх последовательности - это аннотации.
Я бы начала с них: каждый запуск mfold может генерировать группу аннотаций к анализируемой последовательности.

Если усложнять дальше - то можно исходить из того, что каждая и аннотаций является семантически нагруженным элементом, который может иметь свои средства визуализации. Например, на circular view для шпилек показать иконку шпильки рядом с аннотацией.

yalgaer commented:
> Regarding the size of the picture: I agree that changes are required, I wrote them in a separate message, they must be made. But I don’t understand what you mean by "связанными с множеством табличных структур".

Если для отображения графического отчета мы будем использовать внешний браузер, мы можем включить и готовую библиотеку которая сможет манипулировать векторными картинками: зумировать их, обеспечить panning. В подобном подходе я расчитываю на минимум кода с нашей стороны и на максимальное переиспользование готовых решений.

yalgaer commented:
> > 4. Должна быть возможность вернуться к однажды полученному отчету - открыть его заново.  Сегодня если я закрою окно с результатами расчёта я не смогу вернуться к нему снова
> 
> Always save output files, like in the workflow? And the dashboard is like in the workflow?)

Достаточно показывать пользователю папку в файловой системе куда сохраняется отчет. Если далее этот отчет открывается браузером то пользователь будет сразу видеть путь до файлов отчета в браузере и использовать средства браузера чтобы вернуться к отчету: история, закладки.

В самом UGENE в mfold можно не делать никаких изменений.

evelinabis commented:
@yalgaer, идеи в целом понятны. Напиши, пожалуйста, минимальные требования к сдаче.

yalgaer commented:
Думаю минимально будет достаточно сделать следующее

1) В отчете по вычислениям показывать только метаинформацию и ссылку на реальный отчет.
Пример метаинформации: сколько вариантов/элементов нашлось, за какое время и с какими параметрами.

2) Открывать реальный отчет в браузере и немного подтюнить look & feel отчета: добавить заголовочную таблицу со ссылками в начало отчета, добавить минимальные стили к таблицам

Это решит проблемы с подвисанием UGENE и пользователь будет видеть/знать где именно на его диске хранятся отчеты - сможет вернуться к ним.

evelinabis committed:
Add first part of xml tests

evelinabis committed:
Mfold.

Add run info to report.
Show big report in browser only.

evelinabis committed:
Mfold. Fix unicode syms

evelinabis committed:
Mfold. Fix unix build

evelinabis commented:
Обычный отчёт в UGENE
![image](https://github.com/ugeneunipro/ugene/assets/39720063/c2c0a18e-4735-4d32-aad2-4efb24d72376)

Отчёт в браузере
![image](https://github.com/ugeneunipro/ugene/assets/39720063/92cbba1d-5a92-48d3-a6be-c1768f241e60)

Если отчёт большой
![image](https://github.com/ugeneunipro/ugene/assets/39720063/970c89a3-9c37-4de7-8831-ec60fffb9ad6)

И он же в браузере (+установлено высокое разрешение картинки)
![image](https://github.com/ugeneunipro/ugene/assets/39720063/650fdc00-d92b-4a85-8c80-76776adbc07b)
![image](https://github.com/ugeneunipro/ugene/assets/39720063/7818b3ce-6697-4aad-8638-5db8d54a33c6)
![image](https://github.com/ugeneunipro/ugene/assets/39720063/cee8a511-473a-4eeb-aff1-d9a18039badb)

evelinabis committed:
Mfold. Help btn

evelinabis committed:
Merge branch 'master' of https://github.com/ugeneunipro/ugene into mfold1

yalgaer commented:
Пару вопросов/предложений:

1. Что если отчет не большой но их (отчетов) много? Будут ли 10 маленьких отчетов вешать UGENE как 1 большой?

2. Можно ли применить какие-то минимальные стили в таблице открываемом в браузере: чтобы выглядело прилично? Можно  просто добавить минимальный CSS к HTML файлу в заголовок (типа https://picocss.com/docs/table)

3. В загловок отчета можно добавить зависимость от библиотеки которая "бесплатно" дает возможно зумирования и паннинга картинок. Типа этой https://github.com/s-yadav/iv-viewer?tab=readme-ov-file . 
Можно ссылкой на веб ресурс. Если сети не будет - зумирование просто отключится. Можно поддержать и без интернета (вложить JS файл в ресурсы UGENE)- но мне кажется это уже второстепенно.

evelinabis commented:
В процессе....

evelinabis commented:
> * Что если отчет не большой но их (отчетов) много? Будут ли 10 маленьких отчетов вешать UGENE как 1 большой?

Не виснет.

> * Можно ли применить какие-то минимальные стили в таблице открываемом в браузере: чтобы выглядело прилично? Можно  просто добавить минимальный CSS к HTML файлу в заголовок (типа https://picocss.com/docs/table)

Делаю. Надо разделить отчёт для UGENE и отчёт для файла в этом случае, т. к. отчёт для файла полноценный отчёт с <DOCTYPE>, телом и стилями, тогда как юджиновский используется как встраиваемая часть в другой всеобъемлющий хтмл-отчёт. В этом случае стили в середине хтмлины (в начале будет инфа о задаче, затем стили, затем таблица с картинками) будет также выглядеть нелепо, ещё и не ясно, как такое кьюту парсить.

> * В загловок отчета можно добавить зависимость от библиотеки которая "бесплатно" дает возможно зумирования и паннинга картинок. Типа этой https://github.com/s-yadav/iv-viewer?tab=readme-ov-file .
>   Можно ссылкой на веб ресурс. Если сети не будет - зумирование просто отключится. Можно поддержать и без интернета (вложить JS файл в ресурсы UGENE)- но мне кажется это уже второстепенно.

Постараюсь. В переспективе надо форкнуть репозиторий, чтобы если что-то случилось с репо, стили не полетели (надеюсь, форк не похерится, если оригинальный репо удалят).

В ресурсы не успею выложить.

yalgaer commented:
> Делаю. Надо разделить отчёт для UGENE и отчёт для файла в этом случае, т. к. отчёт для файла полноценный отчёт с , телом и стилями, тогда как юджиновский используется как встраиваемая часть в другой всеобъемлющий хтмл-отчёт. В этом случае стили в середине хтмлины (в начале будет инфа о задаче, затем стили, затем таблица с картинками) будет также выглядеть нелепо, ещё и не ясно, как такое кьюту парсить.

Я бы не стала поддерживать внутренний отчет UGENE - достаточно внешнего. Но и не возражаю если будешь делать так. QT просто проигнорирует стили/классы которые будет использовать браузер

> > * В загловок отчета можно добавить зависимость от библиотеки которая "бесплатно" дает возможно зумирования и паннинга картинок. Типа этой https://github.com/s-yadav/iv-viewer?tab=readme-ov-file .
> >   Можно ссылкой на веб ресурс. Если сети не будет - зумирование просто отключится. Можно поддержать и без интернета (вложить JS файл в ресурсы UGENE)- но мне кажется это уже второстепенно.

Репозиторий форкать не надо, достаточно либо использовать ссылку на уже опубликованную библиотеку,
```
<script src="https://unpkg.com/iv-viewer/dist/iv-viewer.js"></script>
<link rel="stylesheet" href="https://unpkg.com/iv-viewer/dist/iv-viewer.css">
```
либо
скачать эти файлы и хостить их он нас (GitHub)

evelinabis committed:
Mfold.
Generate reports for UGENE and file separately.
Thus styles for file HTML have been added.

evelinabis committed:
Merge branch 'master' of https://github.com/ugeneunipro/ugene into mfold1

evelinabis commented:
![image](https://github.com/ugeneunipro/ugene/assets/39720063/bf098b05-93c5-4253-b228-65e7dd81182e)
![image](https://github.com/ugeneunipro/ugene/assets/39720063/76feafbf-249d-46ce-9cf9-1cc78e0669b9)
![image](https://github.com/ugeneunipro/ugene/assets/39720063/849a8158-4ad4-4b11-87df-c52c947e5130)

evelinabis committed:
Mfold. Add zoom script and some minor changes

evelinabis committed:
Mfold. Add js plugin to descr

evelinabis committed:
Mfold. Do not fix file:/// logic with previous logic and features

evelinabis committed:
Mfold. Fix path issues

yalgaer commented on src/corelibs/U2Gui/src/util/RegionSelector.cpp:
```diff
@@ -97,6 +97,10 @@ void RegionSelector::showErrorMessage() {
     }
 }
 
+QPair<QLineEdit*, QLineEdit*> RegionSelector::getLineEdits() const {
```
1. Лучше переделать этот метод на 2 именованных.
2. Публичным методам надо добавить документацию
3. Если для целей валидации достаточно вернуть const QListEdit* - лучше вернуть const.

evelinabis replied:
1. Не нравится QPair? Мне тоже. В структурку с именнованными полями стоит обернуть. Разделять не вижу смысла.

Дольше общаться, поэтому сделаю по пожеланиям

evelinabis replied:
С моей точки зрения -- это быстрое и неправильное решение. Стоит переработать архитектуру RegionSelector-а: вынести общие куски в какие-нибудь классы (типа утилитных), переработать по-новому RegionSelector, RegionSelectorWithExcludedRegion (уже есть похожий дублирующий класс) и похоже реализовать RegionSelectorCustomRegionValidation (для mfold), чтобы не было кододублей и одна стилистика.

yalgaer commented on src/plugins/external_tool_support/images/mfold.png:
Иконка выбивается из цветовой гаммы UGENE и выглядит "тяжелой". Я могу помочь подстроить ее если нужно

evelinabis replied:
Да, помоги, пожалуйста

yalgaer commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,163 @@
+    } else if (endEdit->text().isEmpty()) {
+        err = tr("End position not specified");
+    } else if (!startOk) {
+        err = tr("Start position was overflowed");
```
Тут написано: "начальная позиция была переполнена". Лучше написать более общее: Invalid start position

yalgaer commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,163 @@
+        err = tr("End position was overflowed");
+    } else if (!isCircular && startVal > endVal) {
+        err = tr("Start position cannot be greater than end position");
+    } else if (getRegionLen(startVal, endVal) > 3000) {
```
Магическая константа. Надо выделить в обычную константу и добавить документацию

yalgaer commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.cpp:
```diff
@@ -0,0 +1,163 @@
+        err = tr("Start position cannot be greater than end position");
+    } else if (getRegionLen(startVal, endVal) > 3000) {
+        err = tooLongRegionStr;
+    } else {
```
Что если регион слишком короткий?

evelinabis replied:
Задача отработает с ошибкой, что не найдено структур. Поведение выглядит нормальным для меня

yalgaer commented on src/plugins/external_tool_support/src/mfold/MfoldDialog.h:
```diff
@@ -0,0 +1,74 @@
+#include "MfoldSettings.h"
+#include "ui_MfoldDialog.h"
+
+class QComboBox;
```
Точно ли нужны эти классы тут? В cpp в патче я сразу не нашла использования комбо-бокса.

yalgaer commented on src/plugins/external_tool_support/src/mfold/MfoldSupport.cpp:
```diff
@@ -0,0 +1,125 @@
+    ExternalToolSupportSettings::checkTemporaryDir(os);
+    CHECK_OP(os, );
+
+    auto objView = qobject_cast<GObjectViewAction*>(sender())->getObjectView();
```
Все qobject_cast должны быть защищены SAFE_POINT_NN

yalgaer commented on src/plugins/external_tool_support/src/mfold/MfoldSupport.cpp:
```diff
@@ -0,0 +1,125 @@
+
+    auto objView = qobject_cast<GObjectViewAction*>(sender())->getObjectView();
+    auto seqCtx = qobject_cast<AnnotatedDNAView*>(objView)->getActiveSequenceContext();
+    MfoldDialog dialog(*seqCtx);
```
Диалоги не должны создаваться на стеке - объект может быть удален при выходе из exec. Нужно использовать указатель (см другие диалоги)

evelinabis replied:
У нас какие-то отличия в API от Qt? Есть [пример](https://doc.qt.io/qt-5/qdialog.html#:~:text=Code%20Examples-,A%20modal%20dialog%3A,-void%20EditorWindow%3A%3A), там на стеке вроде

yalgaer replied:
Если приложение закрывается с открытым на стеке диалогом то может быть крэш. Поэтому мы используем только диалоги защищенные умным указателем.

yalgaer reviewed (changes requested):
Нужно добавить хотя бы 1 GUI тест на графическую часть добавленного функционала

evelinabis committed:
Mfold. Review & first test

yalgaer committed:
MFold icons update

yalgaer commented:
Я обновила иконки до более простых и изображающих круговую молекулу с петлями вокруг.
Есть замечания по поводу отчета
1. Если запуск завершается без результата то отчет хуже, чем если бы пользователь запустил mfold без UGENE.
Нужно включать лог mfold в отчет как и команду запуска
![image](https://github.com/ugeneunipro/ugene/assets/17791039/ff14dd7f-13b8-412a-8cfc-10e3fd6462d5)

2.  Когда я открываю отчет то не вижу картинок
![image](https://github.com/ugeneunipro/ugene/assets/17791039/0bafc228-4c9f-4ba6-b3ea-9253a0bcface)
В папке с отчетом картинок действительно нет:
![image](https://github.com/ugeneunipro/ugene/assets/17791039/f25c9926-fe45-49ff-8d4c-0d3e619a62d0)

evelinabis commented:
> 2. Когда я открываю отчет то не вижу картинок
>    ![image](https://private-user-images.githubusercontent.com/17791039/307642923-0bafc228-4c9f-4ba6-b3ea-9253a0bcface.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5MzA4MTQsIm5iZiI6MTcwODkzMDUxNCwicGF0aCI6Ii8xNzc5MTAzOS8zMDc2NDI5MjMtMGJhZmMyMjgtNGM5Zi00YmE2LWIzZWEtOTI1M2EwYmNmYWNlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA2NTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhMmI5MTlhYzg4MWVjNzRhNWJiMDM5MDY0YTAyMTYxYzgwZGJiODFlZGFhNWJmNWQzMzg2NmE5OWQwOWNhNjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0tWMg7MJdJP7NuDzY_zx_o8_duUaL4hpZi508THuwU8)
>    В папке с отчетом картинок действительно нет:
>    ![image](https://private-user-images.githubusercontent.com/17791039/307643205-f25c9926-fe45-49ff-8d4c-0d3e619a62d0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5MzA4MTQsIm5iZiI6MTcwODkzMDUxNCwicGF0aCI6Ii8xNzc5MTAzOS8zMDc2NDMyMDUtZjI1Yzk5MjYtZmU0NS00OWZmLThkNGMtMGQzZTYxOWE2MmQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA2NTUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc0NzQ2MzdhMzAzYzk3YTMzZjUzMWI4ODhlY2E5YTE2YjVjNDkxNDcyNGNmNDA1NWJlODAyMzc4NmNkNTQzYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CjsjtHUPtYvOkoG0T0X_g9FDO_r8-B55rV61Oq5n4hY)

Обнови бинари

evelinabis commented:
Лучше собирать релиз и тестировать там, а не выкачивать дебажные сборки

evelinabis committed:
Mfold. Add user friendly message about failure. Add more GUI tests.

evelinabis committed:
Merge branch 'mfold1' of https://github.com/ugeneunipro/ugene into mfold1

evelinabis committed:
Merge branch 'master' of https://github.com/ugeneunipro/ugene into mfold1

yalgaer commented:
Обновила бинарник mfold, картинки появились, но вот в таком виде:
``` html
<td class="imgcell"><img class="image" src="murine.txt_1.png" data-high-res-src="murine.txt_1.png"
```
В итоге картинка выглядит плохо:
![image](https://github.com/ugeneunipro/ugene/assets/17791039/3febe022-a1b6-43dc-95b6-9246e281895c)


Напиши пожалуйста, когда будет можно тестировать готовую версию
И не забудь добавить в отчет информацию о команде запуска и лог запуска

evelinabis commented:
> Обновила бинарник mfold, картинки появились, но вот в таком виде:
> 
> ```
> <td class="imgcell"><img class="image" src="murine.txt_1.png" data-high-res-src="murine.txt_1.png"

Что не так с HTML?

> В итоге картинка выглядит плохо:

Установи DPI повыше.

> Напиши пожалуйста, когда будет можно тестировать готовую версию И не забудь добавить в отчет информацию о команде запуска и лог запуска

Но зачем? Зачем вообще отчёты тогда? Я же уже написала все настройки для повторного запуска в отчёте. А лог вообще мусор

yalgaer commented:
И не забудь добавить в отчет информацию о команде запуска и лог запуска
> 
> Но зачем? Зачем вообще отчёты тогда? Я же уже написала все настройки для повторного запуска в отчёте. А лог вообще мусор

Детали команды запуска mfold с параметрами нужны в отчете чтобы пользователь видел как именно был получет данный отчет.

Лог mfold нужен так как в нем может содержаться ошибка или детали вычисления. Разработчики mfold решили что этот лог нужен (они могли бы его не выдавать, но решили показывать). То есть UGENE не должен скрывать детали/ошибки вычисления

yalgaer commented:
> > Обновила бинарник mfold, картинки появились, но вот в таком виде:
> > ```
> > <td class="imgcell"><img class="image" src="murine.txt_1.png" data-high-res-src="murine.txt_1.png"
> > ```
> 
> Что не так с HTML?
> 
> > В итоге картинка выглядит плохо:
> 
> Установи DPI повыше.

После запуска MFold я вижу в отчете маленькие (20kb) картинки которые выглядят плохо.

- Если мы генерируем пары картинок (маленькая и большая) то в HTML должны быть 2 имени файла.
- Если мы генерируем только большие картинки - то они должны выглядеть лучше.

Просить пользователя поменять DPI монитора - это перебор.

evelinabis commented:
> И не забудь добавить в отчет информацию о команде запуска и лог запуска
> 
> > Но зачем? Зачем вообще отчёты тогда? Я же уже написала все настройки для повторного запуска в отчёте. А лог вообще мусор
> 
> Детали команды запуска mfold с параметрами нужны в отчете чтобы пользователь видел как именно был получет данный отчет.

Это бесполезная информация, которая засоряет отчёт. Тем не менее, будет сделана.

> Лог mfold нужен так как в нем может содержаться ошибка или детали вычисления. Разработчики mfold решили что этот лог нужен (они могли бы его не выдавать, но решили показывать).

Инструмент достаточно старый, а к тому, что другие разработчики нарешали, мы должны подходить с умом. Хочу отметить, что это в первую очередь инструмент по научной статье, а не готовый поддерживаемый продукт.

> После запуска MFold я вижу в отчете маленькие (20kb) картинки которые выглядят плохо.
> 
> * Если мы генерируем пары картинок (маленькая и большая) то в HTML должны быть 2 имени файла.

Мы генерируем 2 картинки: одни лежат во временной директории для отчёта UGENE, другие даются пользователю. Ему не нужны 2 картинки. Картинки никак не связаны по принципу "большая/маленькая картинка". Картинки вообще никак между собой не связаны.

Сколько килобайт должна быть картинка?

> Просить пользователя поменять DPI монитора - это перебор.

Как можно было додуматься до такого? Настройку в диалоге выстави побольше.

evelinabis commented:
Заодно можешь подсказать, какое DPI оставить по умолчанию, чтобы картинка тебя устраивала ;)

yalgaer commented:
> > Детали команды запуска mfold с параметрами нужны в отчете чтобы пользователь видел как именно был получет данный отчет.
> 
> Это бесполезная информация, которая засоряет отчёт. Тем не менее, будет сделана.

Отлично дай знать как будет сделано.

yalgaer commented:
> Заодно можешь подсказать, какое DPI оставить по умолчанию, чтобы картинка тебя устраивала ;)

Создаваемая картинка содержит координаты и имена нуклеотидов. Они должны быть видны на зуме. У меня получилось увидеть четкие буквы только выставив DPI равный 1000. Размер картинки при этом получился равным 1мб.

Для пользователя установка DPI самостоятельно - это магия.

Вариантом решения может быть генерация маленькой preview картинки для отчета и открытие векторного файла при клике на картинку (PDF или PS). В этом случае потери качества не будет.

Итого, я бы предложила убрать DPI из настроек, генерировать маленькое растерное изображение с DPI=96 и при клике на него в отчете открывать векторное изображение без потери качества.

JS библиотеку с зумом картинок можно убрать: то, как она используется в отчете (зум в маленьком окне без четких границ) - это неудобно.

evelinabis commented:
> Для пользователя установка DPI самостоятельно - это магия.

Почему? 1. Это будет прописано в доке. 2. Это прописано в тултипе (непонятно? можно переписать). 3. Некоторые учёные понимают этот параметр. Наши заказчики, к примеру. Один из них сказал, что для статей ему нужно 300 DPI. 4. Ребятки без понимания могут почитать инструкцию/тултип, поиграться с параметром и это понимание обрести.

> Вариантом решения может быть генерация маленькой preview картинки для отчета и открытие векторного файла при клике на картинку (PDF или PS). В этом случае потери качества не будет.

PS не поддерживается браузерами. PDF -- очевидный выбор.

> Итого, я бы предложила убрать DPI из настроек, генерировать маленькое растерное изображение с DPI=96 и при клике на него в отчете открывать векторное изображение без потери качества.
> 
> JS библиотеку с зумом картинок можно убрать: то, как она используется в отчете (зум в маленьком окне без четких границ) - это неудобно.

Сделаю. На мой взгляд, вывод в картинки может быть полезен. Предложила бы оставить эту опцию. Тот, кому она нужна, воспользуется ею. Возможно, для статьи ему нужна будет именно картинка 300 DPI, а не pdf? Тот, кто не разбирается, будет довольствоваться картинками 96 DPI по умолчанию. На отчёте это не скажется: превью-картинка будет основываться на этой высококачественной картинке с заданным пользователем DPI, но её размер в отчёте будет ограничиваться под размеры экрана (например, как сейчас: высота картинки=85% от экрана).

Что скажешь? Оставим DPI?

yalgaer commented:
На png картинке при увеличении должны быть видны символы нуклеотидов и координаты. Если это видно при 300 DPI - пусть будет 300 DPI. Можешь сделать как посчитаешь нужным, а я после протестирую.

PDF (или PS) нужно положить в папку рядом с картинкой и открывать при клике на ссылку (или картинку) в отчете из браузера. Наличие векторного формата решит все вопросы с качеством изображения. Также, думаю, векторный формат для данного типа изображений будет занимать в разы меньше места и идеально подойдет для статей/постеров.

evelinabis committed:
Mfold. Fix long sequence name. Minor refactoring

evelinabis committed:
Mfold.

Decrease max DPI for Chrome rendering compatibility.
Increase default DPI for better resolution.
Show "No hairpins error" only if no hairpins found.
Fix opening folder containing report after deleting the folder.
Remove zoom script.

evelinabis committed:
Mfold. Dump log & refactoring

evelinabis committed:
Merge branch 'master' of https://github.com/ugeneunipro/ugene into mfold1

evelinabis commented:
@yalgaer, посмотри, пожалуйста, http://ugene-teamcity.unipro.ru:8111/ugene-teamcity/buildConfiguration/ReleaseLinuxPortable/276953?buildTab=artifacts

evelinabis committed:
Mfold. More tests

yalgaer commented:
Последнее замечания и я готова дать approve по зеленым тестам:

В отчете 2 раза указывается путь на html файл. Достаточно указать один раз.

![image](https://github.com/ugeneunipro/ugene/assets/17791039/4365547d-cc71-418b-b861-7295b261a211)

evelinabis committed:
Mfold. Remove double out link. Add test

evelinabis committed:
Merge branch 'master' of https://github.com/ugeneunipro/ugene into mfold1

evelinabis committed:
Mfold. Fix mac coding
